src/mdns/mdns_avahi.cpp

```diff
@@ -860,6 +919,224 @@ void Publisher::Destroy(Publisher *aPublisher)
     delete static_cast<PublisherAvahi *>(aPublisher);
 }
 
+void PublisherAvahi::Subscription::Release(void)
+{
+}
+
+void PublisherAvahi::ServiceSubscription::Browse(void)
+{
+    assert(mPublisherAvahi->mClient != nullptr);
+
+    otbrLogInfo("browse service %s", mType.c_str());
+    if (!avahi_service_browser_new(mPublisherAvahi->mClient, AVAHI_IF_UNSPEC, AVAHI_PROTO_INET6, mType.c_str(),
+                                   mPublisherAvahi->mDomain, static_cast<AvahiLookupFlags>(0), HandleBrowseResult,
+                                   this))
+    {
+        otbrLogWarning("failed to browse service %s: %s", mType.c_str(),
+                       avahi_strerror(avahi_client_errno(mPublisherAvahi->mClient)));
+    }
+}
+
+void PublisherAvahi::ServiceSubscription::HandleBrowseResult(AvahiServiceBrowser *  aServiceBrowser,
+                                                             AvahiIfIndex           aInterfaceIndex,
+                                                             AvahiProtocol          aProtocol,
+                                                             AvahiBrowserEvent      aEvent,
+                                                             const char *           aName,
+                                                             const char *           aType,
+                                                             const char *           aDomain,
+                                                             AvahiLookupResultFlags aFlags,
+                                                             void *                 aContext)
+{
+    static_cast<PublisherAvahi::ServiceSubscription *>(aContext)->HandleBrowseResult(
+        aServiceBrowser, aInterfaceIndex, aProtocol, aEvent, aName, aType, aDomain, aFlags);
+}
+
+void PublisherAvahi::ServiceSubscription::HandleBrowseResult(AvahiServiceBrowser *  aServiceBrowser,
+                                                             AvahiIfIndex           aInterfaceIndex,
+                                                             AvahiProtocol          aProtocol,
+                                                             AvahiBrowserEvent      aEvent,
+                                                             const char *           aName,
+                                                             const char *           aType,
+                                                             const char *           aDomain,
+                                                             AvahiLookupResultFlags aFlags)
+{
+    OTBR_UNUSED_VARIABLE(aProtocol);
+
+    otbrLogInfo("browse service reply: %s.%s%s inf %u, flags=%u", aName, aType, aDomain, aInterfaceIndex, aFlags);
+
+    if (aEvent == AVAHI_BROWSER_FAILURE)
+    {
+        mPublisherAvahi->OnServiceResolveFailed(*this, avahi_client_errno(mPublisherAvahi->mClient));
+    }
+    else
+    {
+        Resolve(aInterfaceIndex, aName, aType, aDomain);
+    }
+    avahi_service_browser_free(aServiceBrowser);
```

Comment:
What if the service is unsubscribed before `HandleBrowseResult` is called for the first time?
Should we free the browser/resolve immediately in Release()?